Importing lodash, rxjs fails in node environment, globals undefined

**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

Importing `lodash@3.10.1` and `rxjs@5.0.0-beta.11` fails in Jest, when `testEnvironment` is set to `node`.

Here's a minimal repository reproduce the bug: https://github.com/fson/sandbox/tree/3ad45f19a1a3efa711cc8eba44fe349673471092/jest-node-environment-repro (just `npm install` and `npm test`).

It seems that the problem is that some globals (e.g. `Array`, `Object`) are missing from the [`global` object created by `jest-environment-node`](https://github.com/facebook/jest/blob/c8a18d4e0fb4eb51fafef6fbdb80b659c8828608/packages/jest-environment-node/src/index.js#L28-L40).

**What is the expected behavior?**

Expected to be able to import `lodash` and `rxjs` without errors thrown.

Tested on OS X 10.11.6, Node.js v6.4.0.

```
> npm test -- --debug

> jest-node-environment-repro@1.0.0 test /Users/ville/Projects/sandbox/jest-node-environment-repro
> jest "--debug"

jest version = 15.1.1
test framework = jasmine2
config = {
  "testEnvironment": "/Users/ville/Projects/sandbox/jest-node-environment-repro/node_modules/jest-environment-node/build/index.js",
  "rootDir": "/Users/ville/Projects/sandbox/jest-node-environment-repro",
  "name": "-Users-ville-Projects-sandbox-jest-node-environment-repro",
  "setupFiles": [
    "/Users/ville/Projects/node_modules/babel-polyfill/lib/index.js"
  ],
  "testRunner": "/Users/ville/Projects/sandbox/jest-node-environment-repro/node_modules/jest-jasmine2/build/index.js",
  "scriptPreprocessor": "/Users/ville/Projects/sandbox/jest-node-environment-repro/node_modules/babel-jest/build/index.js",
  "usesBabelJest": true,
  "preprocessorIgnorePatterns": [
    "/node_modules/"
  ],
  "automock": false,
  "bail": false,
  "browser": false,
  "cacheDirectory": "/var/folders/qm/_psd4jjx3t3d_7hr4w4m6jpr0000gn/T/jest",
  "colors": false,
  "coveragePathIgnorePatterns": [
    "/node_modules/"
  ],
  "coverageReporters": [
    "json",
    "text",
    "lcov",
    "clover"
  ],
  "globals": {},
  "haste": {
    "providesModuleNodeModules": []
  },
  "mocksPattern": "__mocks__",
  "moduleDirectories": [
    "node_modules"
  ],
  "moduleFileExtensions": [
    "js",
    "json",
    "node"
  ],
  "moduleNameMapper": {},
  "modulePathIgnorePatterns": [],
  "noStackTrace": false,
  "notify": false,
  "preset": null,
  "resetModules": false,
  "testPathDirs": [
    "/Users/ville/Projects/sandbox/jest-node-environment-repro"
  ],
  "testPathIgnorePatterns": [
    "/node_modules/"
  ],
  "testRegex": "(/__tests__/.*|\\.(test|spec))\\.js$",
  "testURL": "about:blank",
  "timers": "real",
  "useStderr": false,
  "verbose": null,
  "watch": false,
  "cache": true,
  "watchman": true,
  "testcheckOptions": {
    "times": 100,
    "maxSize": 200
  }
}
 FAIL  ./rxjs.test.js
  ● can import rxjs

    TypeError: Cannot read property 'assign' of undefined

      at Object.<anonymous> (node_modules/rxjs/util/assign.js:4:18)

 FAIL  ./lodash.test.js
  ● can import lodash

    TypeError: Cannot read property 'prototype' of undefined

      at runInContext (node_modules/lodash/index.js:718:27)
      at Object.it (lodash.test.js:2:18)

 FAIL  ./globals.test.js
  ● defines globals like Node.js does

    expect(received).toBeDefined()

    Expected value to be defined, instead received
      "undefined"

      at Object.it (globals.test.js:2:24)

Test Summary
 › Ran all tests.
 › 3 tests failed, 0 tests passed (3 total in 3 test suites, run time 0.649s)
npm ERR! Test failed.  See above for more details.
```

--- a/integration_tests/__tests__/global-test.js
+++ b/integration_tests/__tests__/global-test.js
@@ -1,0 +1,14 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @emails oncall+jsinfra
+ */
+'use strict';
+
+test('globals are properly defined', () => {
+  expect(global.Object).toBe(Object);
+});

--- a/integration_tests/auto-clear-mocks/package.json
+++ b/integration_tests/auto-clear-mocks/package.json
@@ -1,0 +1,5 @@
+{
+  "jest": {
+    "testEnvironment": "node"
+  }
+}

--- a/packages/jest-environment-node/src/index.js
+++ b/packages/jest-environment-node/src/index.js
@@ -18,26 +18,23 @@ const installCommonGlobals = require('jest-util').installCommonGlobals;
 const ModuleMocker = require('jest-mock');
 const vm = require('vm');
 
-const isNaN = global.isNaN;
-
 class NodeEnvironment {
 
+  context: ?vm$Context;
   fakeTimers: ?FakeTimers;
   global: ?Global;
   moduleMocker: ?ModuleMocker;
 
   constructor(config: Config) {
-    const global = this.global = {};
-    vm.createContext(this.global);
+    this.context = vm.createContext();
+    const global = this.global = vm.runInContext('this', this.context);
     global.global = global;
     global.clearInterval = clearInterval;
     global.clearTimeout = clearTimeout;
-    global.setInterval = setInterval;
-    global.setTimeout = setTimeout;
-    global.isNaN = isNaN;
-    global.ArrayBuffer = ArrayBuffer;
     global.JSON = JSON;
     global.Promise = Promise;
+    global.setInterval = setInterval;
+    global.setTimeout = setTimeout;
     installCommonGlobals(global, config.globals);
     this.moduleMocker = new ModuleMocker();
     this.fakeTimers = new FakeTimers(global, this.moduleMocker, config);
@@ -47,13 +44,13 @@ class NodeEnvironment {
     if (this.fakeTimers) {
       this.fakeTimers.dispose();
     }
-    this.global = null;
+    this.context = null;
     this.fakeTimers = null;
   }
 
   runScript(script: Script): ?any {
-    if (this.global) {
-      return script.runInContext(this.global);
+    if (this.context) {
+      return script.runInContext(this.context);
     }
     return null;
   }

--- a/packages/jest-runtime/src/transform.js
+++ b/packages/jest-runtime/src/transform.js
@@ -90,11 +90,11 @@ const writeCacheFile = (cachePath: Path, fileData: string) => {
   }
 };
 
-const wrap = content => '({"' +
-  EVAL_RESULT_VARIABLE +
+const wrap = content =>
+  '({"' + EVAL_RESULT_VARIABLE +
   '":function(module,exports,require,__dirname,__filename,global,jest){' +
-   content +
-   '\n}});';
+  content +
+  '\n}});';
 
 const readCacheFile = (filePath: Path, cachePath: Path): ?string => {
   if (!fileExists(cachePath)) {

